core/libp2p/basic/readable.hpp

```diff
@@ -6,18 +6,44 @@
 #ifndef KAGOME_READABLE_HPP
 #define KAGOME_READABLE_HPP
 
-#include "libp2p/common/network_message.hpp"
+#include <functional>
+#include <outcome/outcome.hpp>
+#include <system_error>
+
+#include "common/buffer.hpp"
 
 namespace libp2p::basic {
 
-class Readable {
-  /**
-     * Read message from the connection
-     * @return optional to message, received by that connection
+  class Readable {
+   public:
+    using BufferResult = outcome::result<kagome::common::Buffer>;
+    using BufferResultCallback = void(BufferResult);
+
+    /**
+     * @brief This function is used to read data. The function call will block
+     * until exactly {@param to_read} bytes of data has been read successfully,
+     * or until an error occurs.
+     * @param to_read number of bytes to read.
+     * @return result of the operation.
+     */
+    virtual BufferResult read(uint32_t to_read) = 0;
+
+    /**
+     * @brief This function is used to read data. The function call will block
+     * until one or more bytes of data has been read successfully, or until an
+     * error occurs.
+     * @param to_read number of bytes to read.
+     * @return result of the operation.
+     */
+    virtual BufferResult readSome(uint32_t to_read) = 0;
+
+    /**
+     * @brief Asynchronously read everything that was sent to our socket.
+     * @param cb Callback that accepts read result.
      */
-  virtual boost::optional<common::NetworkMessage> read() const = 0;
-};
+    virtual void readAsync(std::function<BufferResultCallback> cb) noexcept = 0;
```

Comment:
Shouldn't the function be passed by a constref?

Reply:
This func is moved inside, so does not matter much